### PR TITLE
BERT QAT post-processing support

### PR DESCRIPTION
**Pull request by mgoin**
Added passes to `quantize_qat_export.py` for Transformer-related structures, mainly supporting quantized MatMuls with kernels and biases. 
Also used @bfineran 's new `ONNXGraph` structure to speed up the passes that were taking the longest. With these changes, we can post-process a BERT with 1000+ nodes in just a few seconds on a MacBook Pro

### CODE REVIEW

**Review by bfineran — COMMENTED** [review]
Great job @mgoin - the results look great! Pulled down this branch and ran a BERT qat graph through this, ran in less than 10 seconds. Structure LGTM in netron. There's a few places here we could use the ONNXGraph object more, but given the run time I think it is fine for now. Left a small comment. Will take another pass tomorrow after going through more of the netron graph.

File: src/sparseml/pytorch/utils/quantization/quantize_qat_export.py [hunk 1]
```diff
@@ -508,21 +516,219 @@ def _convert_quantizable_gemm(
...
+            weight_quantize_params.zero_point,
+        )
+        quantized_weight = quantized_weight.transpose()  # Gemm has implicit transpose
+        quantized_weight_name = "{}.weight_quantized".format(gemm_node.name)
+        quantized_weight_initializer = numpy_helper.from_array(
+            quantized_weight, name=quantized_weight_name
+        )
+        model.graph.initializer.append(quantized_weight_initializer)
+
+        """ QLinearMatMul """
```
Comment (bfineran) [inline src/sparseml/pytorch/utils/quantization/quantize_qat_export.py:623]: nit: `#` for inline comments

File: src/sparseml/pytorch/utils/quantization/quantize_qat_export.py [hunk 2]
```diff
@@ -508,21 +516,219 @@ def _convert_quantizable_gemm(
...
+        # create qmatmul node and add it to graph
+        qmatmul_node = onnx.helper.make_node(
+            "QLinearMatMul",
+            qmatmul_inputs,
+            [qmatmul_output],
+            qmatmul_name,
+        )
+        model.graph.node.append(qmatmul_node)
+
+        """ QLinearAdd """
```
Comment (bfineran) [inline src/sparseml/pytorch/utils/quantization/quantize_qat_export.py:648]: nit: `#` for inline comments

File: src/sparseml/pytorch/utils/quantization/quantize_qat_export.py [hunk 3]
```diff
@@ -508,21 +516,219 @@ def _convert_quantizable_gemm(
...
+        qadd_node = onnx.helper.make_node(
+            "QLinearAdd",
+            qadd_inputs,
+            [qadd_output],
+            qadd_name,
+            **kwargs,
+        )
+        model.graph.node.append(qadd_node)
+
+        """ Cleanup """
```
Comment (bfineran) [inline src/sparseml/pytorch/utils/quantization/quantize_qat_export.py:701]: nit: `#` for inline comments

**Review by bfineran — CHANGES_REQUESTED** [review]
That diagram for the qmatmul pass is beautiful. Left a few nits, otherwise looks really good @mgoin. Model in netron looks good too.

**Comment by mgoin** [comment]
thanks @bfineran ! should have addressed your comments now

**Comment by bfineran** [comment]
LGTM pending quality check and the merge conflict!

**Review by kevinaer — DISMISSED** [review]

**Review by bfineran — DISMISSED** [review]

File: src/sparseml/pytorch/utils/quantization/quantize_qat_export.py [hunk 4]
```diff
@@ -508,21 +516,219 @@ def _convert_quantizable_gemm(
         remove_node_and_params_from_graph(model, gemm_node, keep_params=params_to_keep)
 
 
+def _convert_quantizable_matmul_and_add(model: ModelProto):
+    """
+    A pass for converting a MatMul with kernel and bias into a quantized representation
+
+    Starting with:
```
Comment (markurtz) [inline src/sparseml/pytorch/utils/quantization/quantize_qat_export.py:523]: This isn't going to convert well for docs, need to preface each line with | to make sure the formatting is kept exactly as shown in here

File: src/sparseml/pytorch/utils/quantization/quantize_qat_export.py [hunk 5]
```diff
@@ -508,21 +516,219 @@ def _convert_quantizable_gemm(
...
+            |
+        QLinearMatMul (with constant kernel)
+            |
+        QLinearAdd (with constant bias)
+            |
+        DequantizeLinear
+            |
+          OUTPUT
+    """
+    matmul_nodes = [n for n in model.graph.node if n.op_type in ["MatMul"]]
```
Comment (markurtz) [inline src/sparseml/pytorch/utils/quantization/quantize_qat_export.py:553]: could we add some debug logging to this function? seems very long-running and potentially hard to debug

Comment (mgoin) [reply src/sparseml/pytorch/utils/quantization/quantize_qat_export.py:553]: yeah added an info count and debug for each matmul
```
2021-04-19 15:51:55 sparseml.pytorch.utils.quantization.quantize_qat_export INFO     Converted 73 quantizable MatMul ops with weight and bias to QLinearMatMul and QLinearAdd
```

File: src/sparseml/onnx/utils/graph_editor.py [hunk 6]
```diff
@@ -87,8 +87,8 @@ def get_node_parents(
     ) -> List[Union[NodeProto, TensorProto, None]]:
         """
         :param node: node to get the input objects for
-        :return: input nodes or tensors of this node in order. if an input doesn't
-            exist, None will be returned in its place
+        :return: input nodes or tensors of this node in order.
+            if an input doesn't exist, None will be returned in its place
```
[the remark this answers is not in the document]
Comment (markurtz) [reply src/sparseml/onnx/utils/graph_editor.py]: yep, this is unfortunately still an issue. sphinx treats it as the opening of a string. You can escape it you'd like, but I think it's easier to just leave them out

**Review by markurtz — APPROVED** [review]

**Review by bfineran — APPROVED** [review]

**Pull request merged**